### DISCUSSION TITLE
Why are inputs all wrapped in an array?

### DISCUSSION THREAD

**Discussion by ModeratelyGreek** [OpenMDAO/OpenMDAO#3313] [Q&A] [answered]
When implementing compute() on an explicit component, accessing inputs that have been declared to have a simple shape of 1 yields not a python float or int, but an array, wrapping the underlying value. This feels like an odd choice - am I making a mistake? Do I really have to write a few lines of 

my_variable = input["my variable"][0] 

at the start of every class compute method? For all my variables?

This is, regrettably, what the opening lines of compute look like to make variable use easier and avoid nonhomogenous errors creeping up downstream:

```python
def compute(self, inputs, outputs):
        Wing_Leading_Edge_Pos = inputs['Wing Leading Edge Pos'][0]
        Wing_Sweep_Angle = inputs['Wing Sweep Angle'][0]
        Wing_Flap_Length_Percentage = inputs['Wing Flap Length Percentage'][0]
        Wing_Hinge_Chord_Percentage = inputs['Wing Hinge Chord Percentage'][0]
        Wing_Length = inputs['Wing Length'][0]
        Wing_Root_Chord_Length = inputs['Wing Root Chord Length'][0]
        Wing_Tip_Chord_Percentage = inputs['Wing Tip Chord Percentage'][0]
        Wing_Control_Max_Deflection = inputs['Wing Control Max Deflection'][0]
        Tail_Sweep_Angle = inputs['Tail Sweep Angle'][0]
        Tail_Dihedral = inputs['Tail Dihedral'][0]
        Tail_Flap_Length_Percentage = inputs['Tail Flap Length Percentage'][0]
        Tail_Hinge_Chord_Percentage = inputs['Tail Hinge Chord Percentage'][0]
        Tail_Root_Chord_Length = inputs['Tail Root Chord Length'][0]
        Tail_Tip_Chord_Percentage = inputs['Tail Tip Chord Percentage'][0]
        Tail_Control_Max_Deflection = inputs['Tail Control Max Deflection'][0]
        Tail_Aero_Length = inputs['Tail Aero Length'][0]
        Fuselage_Dimensions = inputs['Fuselage Dimensions']
        Airspeed = inputs['Airspeed'][0]
        AoA = inputs['AoA'][0]
        Tail_Rod_Length = inputs['Tail Rod Length'][0]
        Tail_Rod_Radius = inputs['Tail Rod Radius'][0]
```

Any advice? Best practices?

**Comment by robfalck** [comment, marked as answer]
OpenMDAO attempts to do things as efficiently as possible in Python.

One aspect of this is that it maintains contiguous arrays of data for inputs and outputs, so that moving/copying this data can be done more efficiently. This also allows the user to take advantage of things like vectorization through python packages like numpy, scipy, jax, etc.

Without seeing the calculations being performed in compute, its difficult to say, but getting these values as arrays of data shouldn't be a problem if you're using functions from numpy.

With so many variables, there is going to be a good bit of typing for assigning them to local variables, but one way you can cut down on the boilerplate code in this method is to take advantage of the `values` method, which will provide a sequence of the input values in the order in which they were added to the component.

```python
def compute(self, inputs, outputs):
        Wing_Leading_Edge_Pos, Wing_Sweep_Angle, Wing_Flap_Length_Percentage, \
        Wing_Hinge_Chord_Percentage, Wing_Length, Wing_Root_Chord_Length, \
        Wing_Tip_Chord_Percentage, Wing_Control_Max_Deflection, Tail_Sweep_Angle, Tail_Dihedral, \
        Tail_Flap_Length_Percentage, Tail_Hinge_Chord_Percentage, Tail_Root_Chord_Length, \
        Tail_Tip_Chord_Percentage, Tail_Control_Max_Deflection, Tail_Aero_Length, \
        Fuselage_Dimensions,  Airspeed,  AoA,  Tail_Rod_Length, Tail_Rod_Radius = inputs.values()
```

Granted, those variables that are scalar (shape = (1,)) will be length-1 arrays and not python floats, but numpy arrays are generally interchangable in mathematical operations with python floats.

**Reply by ModeratelyGreek** [reply]
Thank you!

I ended up resolving my discomforts with the following line:

```python
inputs = {key: val[0] if isinstance(val, np.ndarray) and val.size == 1 else val for key, val in inputs.items()}
```

**Reply by robfalck** [reply]
Sounds good. Just be careful about overriding the inputs argument. When creating a local variable like that I often would just prepend an underscore (`_inputs = ...`) make sure I'm not messing with mutable arguments.
